### ISSUE TITLE
The scope parameter should be less strict in some of the AcquireTokenXXX methods (allow an empty set of scopes)

### ISSUE THREAD

**Issue by jmprieur** [AzureAD/microsoft-authentication-library-for-dotnet#715]
**Is your feature request related to a problem? Please describe.**
In Web scenarios, the authorization code is requested by the ASP.NET / ASP.NET core middlewre, and redeeming the code is often done with MSAL.NET, but **only** to have a token in the cache, so that it can be used from controllers, using `AcquireTokenSilentAsync`. For instance in the *Web app calls web api* sample

```CSharp
 // This initializes the ASP.NET / ASP.NET Core part
options.ResponseType = "id_token code";
options.Scope.Add("offline_access");
options.Scope.Add("User.Read"); 

// Handling the auth redemption by MSAL.NET so that a token is available in the token cache
// where it will be usable from Controllers later (through the TokenAcquisition service)
var handler = options.Events.OnAuthorizationCodeReceived;
options.Events.OnAuthorizationCodeReceived = async context =>
{
 var _tokenAcquisition = context.HttpContext.RequestServices.GetRequiredService<ITokenAcquisition>();
 await _tokenAcquisition.AddAccountToCacheFromAuthorizationCode(context, options.Scope);
 await handler(context);
 };
```

Note that, as the name of the `AddAccountToCacheFromAuthorizationCode` method tells us, the intent is just to add an account to the cache from the authorization code. For more context see [active-directory-aspnetcore-webapp-openidconnect-v2 in Startup.cs#L67](https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/blob/ee3627031bad6027c69cc79a5ee43af8a9eb4802/Startup.cs#L67)

The simplified implementation of `AddAccountToCacheFromAuthorizationCode` is something like this:

```CSharp
/// <summary>
/// Scopes which are already requested by MSAL.NET. they can't be re-requested;
/// </summary>
private string[] scopesRequestedByMsalNet = new string[] { "openid", "profile", "offline_access" };

var result = await application.AcquireTokenByAuthorizationCodeAsync(context.ProtocolMessage.Code, scopes.Except(scopesRequestedByMsalNet));
```

For more context see [active-directory-aspnetcore-webapp-openidconnect-v2 in Extensions/TokenAcquisition.cs#L71-L130](https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/blob/ee3627031bad6027c69cc79a5ee43af8a9eb4802/Extensions/TokenAcquisition.cs#L71-L130)

The way the token is used in controllers is described [here](https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/blob/ee3627031bad6027c69cc79a5ee43af8a9eb4802/Controllers/HomeController.cs#L41-L64):

```CSharp
 public async Task<IActionResult> Contact()
 {
  var scopes = new string[] { "user.read" };
  try
  {
   var accessToken = await _tokenAcquisition.GetAccessTokenOnBehalfOfUser(HttpContext, scopes);
   dynamic me = await CallGraphApiOnBehalfOfUser(accessToken);
   ViewData["Me"] = me;
   return View();
  }
  catch (MsalUiRequiredException ex)
  {
   if (CanbeSolvedByReSignInUser(ex))
   {
    AuthenticationProperties properties = BuildAuthenticationPropertiesForIncrementalConsent(scopes, ex);
    return Challenge(properties);
   }
   else
   {
    throw;
   }
 }
}
```

`GetAccessTokenOnBehalfOfUser` really calls `AcquireTokenSilentAsync` with the identity of the user signed-in with the application. For details see [here](https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/blob/ee3627031bad6027c69cc79a5ee43af8a9eb4802/Extensions/TokenAcquisition.cs#L309)

**Describe the solution you'd like**
Today, `AcquireTokenByAuthorizationCodeAsync`, 

- does not allow scopes = null
- does not allow scopes = string[0]
- does not allow scopes to contain one of the following strings { "openid", "profile", "offline_access" }

It should not be necessary, at that point to pass-in something in the scope parameter if the web app want to leverage dynamic consent depending on the controllers (given that MSAL.NET will already systematically add "openid profile offline_access" for the token cache to work as expected. It should also be possible to pass one of these without getting an exception.

**Describe alternatives you've considered**
The work around is:
- to avoid passing the known sceopes as  done in TokenAcquisition. It's another thing that customers need to think about, and risks
- pass some random scope ("user.read") to add the token to the cache.

**Additional context**
- The same applies in Web API calling a Web API, and therefore using `AcquireTokenOnBehalfOfAsync` (as illustrated [here](https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/blob/aspnetcore2-2-signInAndCallGraph/Extensions/TokenAcquisition.cs#L347) in order to `AddAccountToCacheFromJwt`):

- For PubliClientApplications, I wonder if we should allow null / empty scopes in the interactive overrides (which would mean, just sign-in a user). allowing "openid", "profile", "offline_access" is less crucial for them as MSAL.NET does all the work and does not need to cooperate with ASP.NET / ASP.NET Core, but this would not hurt.
- This would not be needed for IWA or U/P (but might not hurt either)

**Comment by jmprieur** [comment]
@henrik-me : raised as discussed last week

**Comment by Mephisztoe** [comment]
I am new to MSAL and have been playing around with it using a sample that uses the first preview of MSAL released (1.0.3041...). I implemented a Xamarin Mobile App using MSAL only for signing in and up - now API Access included (yet). The corresponding call to AcquireTokenAsync passed the ClientID as Scope and it worked. 

Now that I am Setting up a base Project on my own using the most up to date Xamarin.Forms as well as the current Version of MSAL, I notice that this doesn't work anymore. Not only that the Parameters that I shall pass to AcquireTokenAsync changed heavily, passing the ClientID as Scope when API Access is not needed (yet) is not supported anymore. Actually I am not able to find out how to get my sample working again using Version 2.7 of MSAL...

**Comment by jennyf19** [comment]
@Mephisztoe You might find this interesting on the [differences between msal v1 and v2](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/wiki/msal-net-2-released#changes-in-msalnet-between-1x-and-2x) as well as [how to migrate from v1 to v2](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/wiki/msal-net-2-released#migrating-from-msalnet-1x-to-msalnet-2x)

Are you working on a b2c app? if so, here is a nice [xamarin sample](https://github.com/Azure-Samples/active-directory-b2c-xamarin-native) that might be helpful.

And a [non-b2c sample with xamarin](https://github.com/Azure-Samples/active-directory-xamarin-native-v2)

Can you send the stack trace or error message you are getting and which platform (Android, iOS, UWP)?

**Comment by Mephisztoe** [comment]
Thanks @jennyf19! I was Looking for information exactly like this (migration notes, etc.). If these details don't help, I will come back to you. I believe my problem is related to some misunderstanding of the concepts behind. For example: Practically any sample shows how to define scopes for getting token to access specific API methods. However, My app doesn't have any API yet and all I want to get working at the moment is using MSAL to sign up a new user (which worked pretty good with a sample that was built on top of the MSAL 1.x). Anway, I will try to get things working with the help of your links. Thank you!!

**Comment by jmprieur** [comment]
What is your scenario, @Mephisztoe 
What do you want to do with the signed-in user if you don't call an API?

Would it be an issue if you just passed new string[] {"user.read"} as a scope?

**Comment by jmprieur** [comment]
@MarkZuber @bgavrilMS @jennyf19 @trwalke 
This is the issue I was mentionning during the grooming meeting 
cc: @henrik-me

**Comment by MarkZuber** [comment]
@jmprieur I'm not clear where the problem is arising here and we can talk in standup on this Friday.  From reading through the code, MSAL does not throw an exception if the list of scopes is null or empty.  There are a few places where we set it if it's not specified such as in ClientApplicationBase:

```
if (commonParameters.Scopes == null || !commonParameters.Scopes.Any())
{
    commonParameters.Scopes = new SortedSet<string>
    {
        ClientId + "/.default"
    };
    requestContext.Logger.Info(LogMessages.NoScopesProvidedForRefreshTokenRequest);
}
```

But we don't do anything in the Validate() methods to ensure that Scopes sent in are not empty.

**Comment by jmprieur** [comment]
@MarkZuber 
Let me check the behavior with MSAL 3.x (this did not work in MSAL 2.x in December)

**Comment by henrik-me** [comment]
@jmprieur : Is this issue still relevant? I believe we support null and empty scopes. not sure if we handle the std. oidc scopes nicely though. Did you get to validate current behavior? Perhaps this can be boiled down to ignoring if oidc scopes are passed in?

**Comment by jmprieur** [comment]
@bgavrilMS : assigning to you, as agreed.

**Comment by bgavrilMS** [comment]
We do not currently support "null or empty" scopes. I believe we handle OIDC scopes in a good way (there are some bugs but I'm fixing them)

**Comment by bgavrilMS** [comment]
Looking into this, it is quite complex, because MSAL assumes there is always an AT returned. But if you request 0 scopes (i.e. "openid", "offline_access" and "profile"), an AT is not returned.

**Comment by bgavrilMS** [comment]
Some initial work for this https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/tree/bogavril/715

**Comment by bgavrilMS** [comment]
A few questions: 

1. For client credentials, I think this is not a valid scenario - i.e. scopes must always be requested
2. For AcquireTokenSilent - how should MSAL behave? If there are 0 tokens, should we return an AuthenticationResult made up of an IdToken only?

**Comment by jmprieur** [comment]
- yes, indeed, @bgavrilMS : for client credentials one needs to profile a "resource/.default" scope
- AcquireTokenSilent yes, I think that IDToken only would be good. An alternative is not allow 0 scope if for AcquireTokenSilent. Maybe that's better.
- Also OBO requires a scope.

**Comment by bgavrilMS** [comment]
@jmprieur - thx for clarifications especially around OBO

I can now confirm that B2C does not issue an access token by default. So this will be some extra work required. 

I think AcquireTokenSilent should allow 0 scopes. Let me see if I can serve the Id Token from the cache.

**Comment by bgavrilMS** [comment]
Just so we're on the same page: 

- AcquireTokenXXX, on B2C, returns an AuthResult that looks like this: 

![image](https://user-images.githubusercontent.com/12273384/78034186-48e0c980-735f-11ea-8a32-6abd1d2e5fbd.png)

ExpiresOn is a struct, so it cannot be null. That value is the default (DateTime.Min). 

- AcquireTokenSilent always perfoms a refresh_token flow
- OBO and ClientCreds do not support this.

**Comment by bgavrilMS** [comment]
I had a long conversation with @shoatman about the behavior of `AcquireTokenSilent` with 0 scopes. `AcquireTokenSilent` is meant to retrieve a fresh access token, not a fresh id token. MSAL was not built to handle id tokens as more than metadata, so parsing the IDToken and fetching the `exp` claim for more advance uses is uncharted territory. 

As such, I've decided to follow @jmprieur 's initial advice to not allow `AcquireTokenSilent` with 0 scopes on B2C. Users can always call `AcquireTokenInteractive` again and re-login if needed - the browser has enough state to fast-track the experience.

**Comment by jmprieur** [comment]
Thanks for confirming, @bgavrilMS and for following-up with @shoatman . This makes sense.

**Comment by neha-bhargava** [comment]
Fixed in [MSAL 4.12.0 release](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/releases/tag/4.12.0)
